### ISSUE TITLE
 New Sessions Expire during Device Setup

### ISSUE THREAD

**Issue by edwardbc** [aws/amazon-chime-sdk-js#400]
**Meeting and attendee IDs:**
N/A

**Timeline**
Outline the main actions the user performed and when:

* First and only user on meeting loads app that receives meeting and attendee info
* User waits more than five minutes in the device selection.
* User tries to join the Meeting after this wait

Since the user took more than 5 minutes without adding audio to the newly created meeting, it expires (by default Chime rules) and there's a WebSocket error and a loop of constant errors afterwards.

**Describe the bug**
There doesn't seem to be an event thrown for this scenario on `AudioVideoObserver` to handle this and also the SDK doesn't provide way to avoid the five minute expiration in this scenario (which is quite common, as many Meeting organizers tend to leave setup screens open minutes before starting). 

**Error Log** (on Meeting start after user waited more than five minutes in device selection)
```
2020-05-20T20:24:58.478Z [INFO] SDK - API/DefaultAudioVideoFacade/45d90abf-7532-4d34-9df4-180111b86338/492f18d8-fe05-47fc-983a-b8ba5a9d07a5/stopVideoPreviewForVideoInput "video-preview"
ConsoleLogger.ts:69 2020-05-20T20:24:58.478Z [INFO] SDK - video send has ideal max bandwidth 1400 kbps
ConsoleLogger.ts:69 2020-05-20T20:24:58.479Z [INFO] SDK - API/DefaultAudioVideoFacade/45d90abf-7532-4d34-9df4-180111b86338/492f18d8-fe05-47fc-983a-b8ba5a9d07a5/chooseVideoInputDevice null
ConsoleLogger.ts:69 2020-05-20T20:24:58.479Z [INFO] SDK - serial group task AudioVideoStart/45d90abf-7532-4d34-9df4-180111b86338/492f18d8-fe05-47fc-983a-b8ba5a9d07a5 completed subtask AudioVideoStart/45d90abf-7532-4d34-9df4-180111b86338/492f18d8-fe05-47fc-983a-b8ba5a9d07a5/MonitorTask
ConsoleLogger.ts:69 2020-05-20T20:24:58.479Z [INFO] SDK - serial group task AudioVideoStart/45d90abf-7532-4d34-9df4-180111b86338/492f18d8-fe05-47fc-983a-b8ba5a9d07a5 running subtask AudioVideoStart/45d90abf-7532-4d34-9df4-180111b86338/492f18d8-fe05-47fc-983a-b8ba5a9d07a5/ReceiveAudioInputTask
ConsoleLogger.ts:69 2020-05-20T20:24:58.479Z [INFO] SDK - running task AudioVideoStart/45d90abf-7532-4d34-9df4-180111b86338/492f18d8-fe05-47fc-983a-b8ba5a9d07a5/ReceiveAudioInputTask
ConsoleLogger.ts:69 2020-05-20T20:24:58.480Z [INFO] SDK - checking whether existing audio device can be reused
ConsoleLogger.ts:69 2020-05-20T20:24:58.480Z [INFO] SDK - reusing existing audio device
ConsoleLogger.ts:69 2020-05-20T20:24:58.481Z [INFO] SDK - API/DefaultDeviceController/chooseVideoInputDevice null -> "PermissionGrantedByBrowser"
ConsoleLogger.ts:69 2020-05-20T20:24:58.481Z [INFO] SDK - using audio device label=Default - MacBook Pro Microphone (Built-in) id=9c190892-4ed7-4df6-9046-8254993aae7f
ConsoleLogger.ts:69 2020-05-20T20:24:58.481Z [INFO] SDK - AudioVideoStart/45d90abf-7532-4d34-9df4-180111b86338/492f18d8-fe05-47fc-983a-b8ba5a9d07a5/ReceiveAudioInputTask took 2 ms
ConsoleLogger.ts:69 2020-05-20T20:24:58.482Z [INFO] SDK - serial group task AudioVideoStart/45d90abf-7532-4d34-9df4-180111b86338/492f18d8-fe05-47fc-983a-b8ba5a9d07a5 completed subtask AudioVideoStart/45d90abf-7532-4d34-9df4-180111b86338/492f18d8-fe05-47fc-983a-b8ba5a9d07a5/ReceiveAudioInputTask
ConsoleLogger.ts:69 2020-05-20T20:24:58.482Z [INFO] SDK - serial group task AudioVideoStart/45d90abf-7532-4d34-9df4-180111b86338/492f18d8-fe05-47fc-983a-b8ba5a9d07a5 running subtask AudioVideoStart/45d90abf-7532-4d34-9df4-180111b86338/492f18d8-fe05-47fc-983a-b8ba5a9d07a5/Timeout15000ms
ConsoleLogger.ts:69 2020-05-20T20:24:58.482Z [INFO] SDK - running task AudioVideoStart/45d90abf-7532-4d34-9df4-180111b86338/492f18d8-fe05-47fc-983a-b8ba5a9d07a5/Timeout15000ms
ConsoleLogger.ts:69 2020-05-20T20:24:58.482Z [INFO] SDK - running task AudioVideoStart/45d90abf-7532-4d34-9df4-180111b86338/492f18d8-fe05-47fc-983a-b8ba5a9d07a5/Timeout15000ms/Media
ConsoleLogger.ts:69 2020-05-20T20:24:58.482Z [INFO] SDK - serial group task AudioVideoStart/45d90abf-7532-4d34-9df4-180111b86338/492f18d8-fe05-47fc-983a-b8ba5a9d07a5/Timeout15000ms/Media running subtask AudioVideoStart/45d90abf-7532-4d34-9df4-180111b86338/492f18d8-fe05-47fc-983a-b8ba5a9d07a5/Timeout15000ms/Media/Setup
ConsoleLogger.ts:69 2020-05-20T20:24:58.483Z [INFO] SDK - running task AudioVideoStart/45d90abf-7532-4d34-9df4-180111b86338/492f18d8-fe05-47fc-983a-b8ba5a9d07a5/Timeout15000ms/Media/Setup
ConsoleLogger.ts:69 2020-05-20T20:24:58.483Z [INFO] SDK - parallel group task AudioVideoStart/45d90abf-7532-4d34-9df4-180111b86338/492f18d8-fe05-47fc-983a-b8ba5a9d07a5/Timeout15000ms/Media/Setup running subtask AudioVideoStart/45d90abf-7532-4d34-9df4-180111b86338/492f18d8-fe05-47fc-983a-b8ba5a9d07a5/Timeout15000ms/Media/Setup/ReceiveTURNCredentialsTask
ConsoleLogger.ts:69 2020-05-20T20:24:58.484Z [INFO] SDK - running task AudioVideoStart/45d90abf-7532-4d34-9df4-180111b86338/492f18d8-fe05-47fc-983a-b8ba5a9d07a5/Timeout15000ms/Media/Setup/ReceiveTURNCredentialsTask
ConsoleLogger.ts:69 2020-05-20T20:24:58.484Z [INFO] SDK - requesting TURN credentials from https://ccp.cp.ue1.app.chime.aws/v2/turn_sessions
ConsoleLogger.ts:69 2020-05-20T20:24:58.486Z [INFO] SDK - parallel group task AudioVideoStart/45d90abf-7532-4d34-9df4-180111b86338/492f18d8-fe05-47fc-983a-b8ba5a9d07a5/Timeout15000ms/Media/Setup running subtask AudioVideoStart/45d90abf-7532-4d34-9df4-180111b86338/492f18d8-fe05-47fc-983a-b8ba5a9d07a5/Timeout15000ms/Media/Setup/Signaling
ConsoleLogger.ts:69 2020-05-20T20:24:58.487Z [INFO] SDK - running task AudioVideoStart/45d90abf-7532-4d34-9df4-180111b86338/492f18d8-fe05-47fc-983a-b8ba5a9d07a5/Timeout15000ms/Media/Setup/Signaling
ConsoleLogger.ts:69 2020-05-20T20:24:58.487Z [INFO] SDK - serial group task AudioVideoStart/45d90abf-7532-4d34-9df4-180111b86338/492f18d8-fe05-47fc-983a-b8ba5a9d07a5/Timeout15000ms/Media/Setup/Signaling running subtask AudioVideoStart/45d90abf-7532-4d34-9df4-180111b86338/492f18d8-fe05-47fc-983a-b8ba5a9d07a5/Timeout15000ms/Media/Setup/Signaling/OpenSignalingConnectionTask
ConsoleLogger.ts:69 2020-05-20T20:24:58.487Z [INFO] SDK - running task AudioVideoStart/45d90abf-7532-4d34-9df4-180111b86338/492f18d8-fe05-47fc-983a-b8ba5a9d07a5/Timeout15000ms/Media/Setup/Signaling/OpenSignalingConnectionTask
ConsoleLogger.ts:69 2020-05-20T20:24:58.488Z [INFO] SDK - adding connection request to queue: wss://signal.m3.ue1.app.chime.aws/control/45d90abf-7532-4d34-9df4-180111b86338?X-Chime-Control-Protocol-Version=3
ConsoleLogger.ts:69 2020-05-20T20:24:58.488Z [INFO] SDK - no existing connection needs closing
ConsoleLogger.ts:69 2020-05-20T20:24:58.489Z [INFO] SDK - opening connection to wss://signal.m3.ue1.app.chime.aws/control/45d90abf-7532-4d34-9df4-180111b86338?X-Chime-Control-Protocol-Version=3
ConsoleLogger.ts:69 2020-05-20T20:24:58.490Z [INFO] SDK - notifying event: WebSocketConnecting
meetingV2.ts:1559 [DEMO] session connecting. reconnecting: false
DefaultWebSocketAdapter.ts:12 WebSocket connection to 'wss://signal.m3.ue1.app.chime.aws/control/45d90abf-7532-4d34-9df4-180111b86338?X-Chime-Control-Protocol-Version=3&X-Amzn-Version=1.6.9&X-Amzn-User-Agent=chrome-81' failed: Connection closed before receiving a handshake response

... 
...
...
```

Can be replicated on Browser Demo by just waiting five minutes in the device selection screen and then hitting join, on a new meeting.

**Platform**
 - OS: N/A
 - Browser: N/A
 - Browser Version: N/A
 - Amazon Chime SDK GitHub Commit: `3a10d4a15c4678ad0f70b4450feba476a35c61a0`

**Comment by ltrung** [comment]
In the case where we have setup screen on a long time before when the meeting actually starts, could we move the logic to after the setup screen?

**Comment by edwardbc** [comment]
Yes, but I still think the SDK should at least execute an event so clients can handle all types of Session expiration correctly, as some would prefer or need to fetch Meeting data beforehand for various reasons.

**Comment by ltrung** [comment]
Got it. Right now you can setup Event Bridge to get notification if the meeting ends https://docs.aws.amazon.com/chime/latest/ag/automating-chime-with-cloudwatch-events.html.

**Comment by krokhale** [comment]
Experiencing this as well. 5 minutes is too little. Also, If I understand this right, even if a single person goes beyond the 5 minutes, does this invalidate the whole meeting? Should't the meeting work for other attendees regardless of a single person's device not being configured?

Can we specify no expiry time for a meeting?

**Comment by ltrung** [comment]
@krokhale The meeting only automatically expires if no attendee joins in 5 minutes. Did you see the behavior differently?

**Comment by krokhale** [comment]
@ltrung What if the first attendee joins after 5 minutes. Happens all the time. Would I then have to create a new meeting on the fly?

If this is the case, then the onus of maintaining a meeting/session is falling entirely on the application. So an application meeting instance would create a chime meeting instance and have to take care of all the possibilities and edge cases of the 5 minute rule.

Just trying to understand, Why have the 5 minute rule anyway? Why not have an option to set a chime meeting expiry time? Why connect the meeting expiry with an attendee joining a meeting?

A meeting to my mind is just a place where a certain set of people can gather, the onus of expiring the meeting or disqualifying an attendee should lie with the application. Whilst Chime just provides the API's necessary to conduct the meeting.

**Comment by aburnson** [comment]
@krokhale a meeting in the Chime SDK is essentially a reservation to a particular media placement group, which is the collection of endpoints that attendees use to rendezvous to the same media backends. The auto-end policy is designed to clean up meetings which are no longer being used to free up those resources for new meetings and reduce the likelihood that a meeting would be billed for longer than it actually being used.

From the ["Creating meetings..." documentation](https://docs.aws.amazon.com/chime/latest/dg/mtgs-sdk-mtgs.html), a meeting will be auto-ended if:
- No audio connections are present in the meeting for more than five minutes.
- Fewer than two audio connections are present in the meeting for more than 30 minutes.
- The meeting time exceeds 24 hours.

In other words, once a meeting is created there is a 5-minute limit for the first attendee to join and 30-minute limit for 2 or more attendees to join; otherwise, the meeting will be cleaned up. Currently these limits cannot be adjusted, but we have taken your feedback requesting control over these limits.

Our guidance is to create a meeting when you expect the first attendee to join. Any pre-meeting settings dependent on user input which could take an arbitrary amount of time should be done before creating the meeting (if the meeting hasn't already been created and joined by another user).

**Comment by edwardbc** [comment]
Is it possible then to have the SDK to trigger an event when the connection fails at the beginning either due session expiration or any other reasons? 

I think several other scenarios could be benefited from this.

**Comment by krokhale** [comment]
@aburnson I think the The auto-end policy would need to be more intelligent than the conditions that you list above. In real world scenarios, those conditions happen all the time and it would end the meeting for everyone. 

- Suppose the first user joins an application meeting (not chime meeting), once on the meeting page the application creates a chime meeting instance on the fly. Now the user has an issue with his Audio setup, takes longer than 5 and the second user joins on the 6th minute, application checks if a chime meeting instance already exists, it does! But the second user would find that the meeting has already ended. What do you do now? Create a new Chime meeting instance? If yes, then the first user who is trying to connect to the earlier meeting needs to be updated with a new meeting instance and so on.

- Now that's just one example, real time communication anyway has all sorts of edge cases, the 5 minute and 30 minute limits add an extra layer of complexity that does not really add any value to the end user application. It's more of Chime trying to manage it's resources which in my opinion should be invisible to the application.

Anyway as you say these cannot be changed, but yes control over them would be nice.

**Comment by aburnson** [comment]
> Now the user has an issue with his Audio setup, takes longer than 5 and the second user joins on the 6th minute

@krokhale the idea is that you would find/create the meeting after the audio/video devices are set up. As detailed in the [API Overview](https://aws.github.io/amazon-chime-sdk-js/modules/apioverview.html), you can create a DefaultDeviceController and set up devices with that controller before attaching the device controller to the MeetingSession and joining.

**Comment by aburnson** [comment]
> Is it possible then to have the SDK to trigger an event when the connection fails at the beginning either due session expiration or any other reasons?

@edwardbc yes, in the AudioVideoObserver [audioVideoDidStop](https://aws.github.io/amazon-chime-sdk-js/interfaces/audiovideoobserver.html#audiovideodidstop) handler you can look at the [MeetingSessionStatus](https://aws.github.io/amazon-chime-sdk-js/classes/meetingsessionstatus.html) and determine the cause of the stop using the isTerminal, isFailure, and statusCode properties.

**Comment by edwardbc** [comment]
@aburnson that's precisely the issue. On the first five minute expiration, the last event that gets triggered is `audioVideoDidStartConnecting`, but `audioVideoDidStop` never gets called but instead the WebSocket connection error is seen at last.

**Comment by aburnson** [comment]
> @aburnson that's precisely the issue. On the first five minute expiration, the last event that gets triggered is `audioVideoDidStartConnecting`, but `audioVideoDidStop` never gets called but instead the WebSocket connection error is seen at last.

I believe this is related to issues addressed recently by #396 and #415

**Comment by aburnson** [comment]
Closing this issue as the 5 minute auto-end if no attendee joins is an intentional limitation of the SDK designed to optimize backend resources towards active meetings and prevent unintentional resource consumption leading to billing. The main takeaway here is to defer creating the meeting until just before the attendee needs to join (e.g. call `meetingSession.audioVideo.start()`). The 5-minute grace period is intended to cover delays between the client app and the server app and brief losses of network connectivity. It is not intended that this time be used to wait on user input which could take arbitrarily long or be canceled (such as dedicated device setup screens).

That being said, we have taken the feedback about potentially being able to adjust the auto-end policy to make it more flexible.

**Comment by xuesichao** [comment]
There's an internal flag to change this on service side. This is not public to builders and you need to open an AWS support case to request this change.
